signals/src/signal.rs

```diff
@@ -450,6 +458,109 @@ impl<A> Signal for Flatten<A>
     }
 }
 
+#[doc(hidden)]
+pub struct Broadcast<A: Signal>
+    where <A as Signal>::Item: Clone {
+
+    // TODO: should these be Arcs ??
+
+    // The state of this instance
+    state: Rc<RefCell<Option<<A as Signal>::Item>>>,
+
+    // The state of the whole broadcast
+    inner: Rc<RefCell<BroadcastInner<A>>>
+}
+
+impl<A: Signal> Broadcast<A>
+    where <A as Signal>::Item: Clone {
+    fn new(signal: A) -> Self {
+        let state = Rc::new(RefCell::new(None));
+        let inner = BroadcastInner {
+            signal: signal,
+            children: vec![Rc::downgrade(&state)]
+        };
+
+        Self {
+            state: state,
+            inner: Rc::new(RefCell::new(inner))
+        }
+    }
+}
+
+struct BroadcastInner<A: Signal> {
+    children: Vec<Weak<RefCell<Option<<A as Signal>::Item>>>>,
+    signal: A
+}
+
+impl<A: Signal> Clone for Broadcast<A>
+    where <A as Signal>::Item: Clone {
+    fn clone(&self) -> Self {
+        let new_state = Rc::new(RefCell::new(self.state.borrow().clone()));
+        self.inner.borrow_mut().children.push(Rc::downgrade(&new_state));
+
+        Self {
+            state: new_state,
+            inner: Rc::clone(&self.inner)
+        }
+    }
+}
+
+impl<A: Signal> Signal for Broadcast<A>
+    where <A as Signal>::Item: Clone  {
+    type Item = <A as Signal>::Item;
+
+    fn poll(&mut self) -> State<Self::Item> {
+        let mut inner = self.inner.borrow_mut();
+
+        match inner.signal.poll() {
+            State::Changed (value) => {
+
+                // Need to broadcast through the whole lot of children.
+                //
+                // We'll also take this opportunity to clean up any
+                // children that have been dropped.
+                //
+                // Claim: The polling on the inner signal will mean that all
+                // polled children will be woken up when the inner is set,
+                // so we don't need tasks
```

Comment:
I'm 99% sure this claim isn't true. It's possible to spawn on multiple Tasks:

```rust
let broadcaster = signal.broadcast();

PromiseFuture::spawn(
    broadcaster.clone().for_each(...)
);

PromiseFuture::spawn(
    broadcaster.for_each(...)
);
```

This isn't just hypothetical, every time you use `.dynamic()` with `html!` it will run a new `PromiseFuture::spawn`:

```rust
let broadcaster = signal.broadcast();

html!("div", {
    attribute("foo", broadcaster.clone().dynamic());
    attribute("bar", broadcaster.dynamic());
})
```

It will run `PromiseFuture::spawn` twice: once for `foo` and once for `bar`

Reply:
Completely agree; I don't know what I was thinking earlier!